pytorch_lightning/utilities/trainer_cli.py

```diff
@@ -0,0 +1,223 @@
+
+        Example, first implement the trainer.py tool as::
+
+            from mymodels import MyModel
+            from pytorch_lightning.utilities.jsonargparse_utils import LightningCLI
+            LightningCLI(MyModel)
+
+        Then in a shell, run the tool with the desired configuration::
+
+            $ python trainer.py --print-config > config.yaml
+            $ nano config.yaml  # modify the config as desired
+            $ python trainer.py --cfg config.yaml
+
+        Args:
+            model_class: The LightningModule class to train on.
+            datamodule_class: An optional LightningDataModule class.
+            save_config_callback: A callback class to save the training config.
+            trainer_class: An optional extension of the Trainer class.
+            description: Description of the tool shown when running --help.
+            parse_env: Whether environment variables are also parsed.
+            **kwargs: Additional arguments to instantiate Trainer.
+        """
+        if 'callbacks' not in kwargs:
+            kwargs['callbacks'] = []
+
+        self.model_class = model_class
+        self.datamodule_class = datamodule_class
+        self.save_config_callback = save_config_callback
+        self.trainer_class = trainer_class
+        self.trainer_kwargs = kwargs
+
+        self.init_parser(description, parse_env)
+        self.add_arguments_to_parser(self.parser)
+        self.add_core_arguments_to_parser()
+        self.parse_arguments()
+        self.instantiate_classes()
+        self.run()
+
+    def init_parser(
+        self,
+        description: str,
+        parse_env: bool
+    ):
+        """Method that instantiates the argument parser
+
+        Args:
+            description: Description of the tool shown when running --help.
+            parse_env: Whether environment variables are also parsed.
+        """
+        self.parser = LightningArgumentParser(
+            description=description,
+            print_config='--print_config',
+            default_env=parse_env,
+            env_prefix='PL'
+        )
+
+    def add_arguments_to_parser(
+        self,
+        parser: LightningArgumentParser
+    ):
+        """Implement to add extra arguments to parser
+
+        Args:
+            parser: The argument parser object to which arguments should be added
+        """
+        pass
+
+    def add_core_arguments_to_parser(self):
+        """Adds arguments from the core classes to the parser"""
+        self.parser.add_trainer_args(self.trainer_class, 'trainer')
+        self.parser.add_module_args(self.model_class, 'model')
+        if self.datamodule_class is not None:
+            self.parser.add_datamodule_args(self.datamodule_class, 'data')
+
+    def parse_arguments(self):
+        """Parses command line arguments and stores it in self.config"""
+        self.config = self.parser.parse_args()
+
+    def instantiate_classes(self):
```

Comment:
I have a situation similar to [this](https://pytorch-lightning.readthedocs.io/en/latest/datamodules.html#using-a-datamodule), where the model has args that are dependent on the DataModule. Not only that, but the DataModule needs to run the prepare_data and setup methods to have the needed parameters ready. 

I guess that can be addressed by inheriting and overriding the instantiate_classes method?

Reply:
Yes, this could be addressed by overriding instantiate_classes. But maybe we should split instantiate_classes into three methods, so that only the relevant part needs to be modified.

Reply:
Splitting can be a good solution, but it's also important to instantiate in the right order. I'm not sure if the DM, Model, Trainer order is always the right one, but it is the one I had occasion to use.

Reply:
There are now individual `instantiate_{model,datamodule,trainer}` methods. `instantiate_classes` is still there which could be overridden to change order or do things in between.

Reply:
I see intantiate_{model,trainer}, but datamodule seems to be part of prepare_fit_kwargs.

In fact, reading the latest code made me realize that perhaps the cleanest solution would be to move the model instantiation into perpare_fit_kwargs (same as datamodule), and have the user overwrite prepare_fit_kwargs as needed. 

I think that the distinction between initializing the trainer and preparing its kwargs (initializing datamodule/model) is a logical one. Handling any special logic of the datamodule/model interdependency can be encapsulated by asking the user to provide the trainer kwargs. 

Does that make sense to you at all?

Reply:
I changed things a bit to be more consistent. There is still prepare_fit_kwargs but now there is an independent instantiate_datamodule method. Right now the methods are rather small which I did so that if someone wants to extend some method, then the change can be kept small. If it looks too fine grained, certainly parts could be moved to perpare_fit_kwargs. Though note that instantiate methods have changed a bit since I implemented a new mode so that a single cli can support multiple models and datamodules.